test/e2e/lib/metrics.go

```diff
@@ -0,0 +1,201 @@
+// metrics.
+//
+// The hope is that this will allow better understanding of where events are lost. E.g. did the
+// source try to send the event to the channel/broker/sink?
+func printAllPodMetricsIfTestFailed(client *Client) {
+	if !client.T.Failed() {
+		// No failure, so no need for logs!
+		return
+	}
+	pods, err := client.Core.Kube.Kube.CoreV1().Pods(client.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		client.T.Logf("Unable to list pods: %v", err)
+		return
+	}
+	for _, pod := range pods.Items {
+		printPodMetrics(client, pod)
+	}
+}
+
+// printPodMetrics attempts to print the metrics from a single Pod to the test logs.
+func printPodMetrics(client *Client, pod corev1.Pod) {
+	podName := types.NamespacedName{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	}
+
+	metricsPort := -1
+	for _, c := range pod.Spec.Containers {
+		for _, p := range c.Ports {
+			if p.Name == "metrics" {
+				metricsPort = int(p.ContainerPort)
+				break
+			}
+		}
+	}
+	if metricsPort < 0 {
+		client.T.Logf("Pod '%v' does not have a metrics port", podName)
+		return
+	}
+
+	root, err := getRootOwnerOfPod(client, pod)
+	if err != nil {
+		client.T.Logf("Unable to get root owner of the Pod '%v': %v", podName, err)
+		root = "root-unknown"
+	}
+
+	podList := &corev1.PodList{
+		Items: []corev1.Pod{
+			pod,
+		},
+	}
+	localPort, err := findAvailablePort()
+	if err != nil {
+		client.T.Logf("Unable to find an avaiable port for Pod '%v': %v", podName, err)
+		return
+	}
+	// There is almost certainly a better way to do this, but for now, just use kubectl to port
+	// forward and use HTTP to read the metrics.
+	pid, err := monitoring.PortForward(client.T.Logf, podList, localPort, metricsPort, pod.Namespace)
+	if err != nil {
+		client.T.Logf("Unable to port forward for Pod '%v': %v", podName, err)
+		return
+	}
+	defer monitoring.Cleanup(pid)
+
+	// Port forwarding takes a bit of time to start running, so try GETs until it works.
+	var resp *http.Response
+	err = wait.PollImmediate(500*time.Millisecond, 10*time.Second, func() (bool, error) {
+		req, _ := http.NewRequest("GET", fmt.Sprintf("http://localhost:%v/metrics", localPort), nil)
+		req.Close = true
+		c := &http.Client{
+			Transport: &http.Transport{DisableKeepAlives: true},
+		}
+		var innerErr error
+		resp, innerErr = c.Do(req)
+		if net.IsConnectionRefused(innerErr) {
+			return false, nil
+		} else {
+			return true, innerErr
+		}
+	})
+
+	if err != nil {
+		client.T.Logf("Unable to read metrics from Pod '%v' (root %q): %v", podName, root, err)
+		return
+	}
+	defer resp.Body.Close()
+	if resp.ContentLength == 0 {
+		client.T.Logf("Pod had no metrics reported '%v' (root %q)", podName, root)
+		return
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		client.T.Logf("Unable to read HTTP response body for root %q: %v", root, err)
+		return
+	}
+	client.T.Logf("Metrics logs for root %q: %s", root, string(b))
```

Comment:
I feel it not quite easy to find these logs as `build-log.txt` is usually a big file.
You can add a new regex in Prow [Spyglass config](https://github.com/knative/test-infra/blob/7a28f47cdbcb11b92359528689b57030fafc127d/config/prod/prow/core/config.yaml#L62-L84) and use it to highlight the metric logs here. But it's up to you.

Reply:
https://github.com/knative/test-infra/pull/2082

I'll leave it as a draft PR until this one merges.